### PR TITLE
Checking for overdue items. [TFF-78]

### PR DESCRIPTION
**Pull request by Cafferoq**
### Summary

Changed student store to hold items with students instead of just addresses. Added timestamps for due time when an item gets checked out. Checks that the students don't have overdue items before allowing a successful checkout.

Note: This PR has a sister PR on the consus-client repo. They should be merged together or not at all.

### Checklist

- [x] Did you run `npm test`?
- [x] Did you update/add documentation for new methods or changed functionality?

Finally, is it ready to be reviewed and merged: yes :white_check_mark: 

### How to Review

1. Check that new functionality doesn't interfere with anything. 
2. Ensure added functionality makes sense. 

### Links

Put links to Jira tickets, and anything else, here.

- [TFF-78](https://msoese.atlassian.net/browse/TFF-78)

### CODE REVIEW

File: src/store/item-store.js [hunk 1]
```diff
@@ -51,6 +51,11 @@ store.registerHandler('NEW_ITEM', data => {
 store.registerHandler('NEW_CHECKOUT', data => {
     data.itemAddresses.forEach(address => {
         store.getItemByAddress(address).status = 'CHECKED_OUT';
+        
+        let timestamp = new Date();
+        timestamp.setUTCHours(16);//Set the time to 5pm today;
```
Comment (tondef001) [inline src/store/item-store.js]: Any particular reason it's set to 5?

Comment (Cafferoq) [inline src/store/item-store.js]: Nope. Just figured that was a good tentative time for now. It'll have to be set-able for long term checkouts anyway.

Comment (tondef001) [inline src/store/item-store.js]: 👍

File: src/store/student-store.js [hunk 2]
```diff
@@ -1,10 +1,11 @@
 import { Store } from 'consus-core/flux';
+import ItemStore from './item-store';
 
 let students = {};
 students[123456] = {
     id: 123456,
     name: 'John von Neumann',
-    itemAddresses: []
+    items: []
```
Comment (thofmann) [inline src/store/student-store.js]: :clap:

File: src/store/checkout-store.js [hunk 3]
```diff
@@ -18,8 +19,13 @@ store.registerHandler('NEW_CHECKOUT', data => {
         studentId: data.studentId,
         itemAddresses: data.itemAddresses
     };
-    checkoutsByActionId[data.actionId] = checkout;
-    checkouts.push(checkout);
+
+    if (StudentStore.hasOverdueItem(data.studentId)){
+        throw new Error('Student has overdue item');
```
Comment (thofmann) [inline src/store/checkout-store.js]: What happens when we throw an error in the action handler?

File: src/store/student-store.js [hunk 4]
```diff
@@ -18,6 +19,12 @@ class StudentStore extends Store {
         return studentsByActionId[actionId];
     }
 
+    hasOverdueItem(id){
+        return students[id].items.some(item => {
```
Comment (thofmann) [inline src/store/student-store.js]: TIL: `Array.prototype.some`

Neat, thanks.

File: src/store/student-store.js [hunk 5]
```diff
@@ -26,15 +33,18 @@ store.registerHandler('NEW_STUDENT', data => {
...
     studentsByActionId[data.actionId] = student;
     students.push(student);
 });
 
 store.registerHandler('NEW_CHECKOUT', data => {
     let student = store.getStudentById(data.studentId);
-    student.itemAddresses = student.itemAddresses.concat(data.itemAddresses);
+
+    data.itemAddresses.forEach(itemAddress => {
+        student.items.push(ItemStore.getItemByAddress(itemAddress));
```
Comment (thofmann) [inline src/store/student-store.js]: :+1:

File: src/store/item-store.js (hunk 1 above)
Comment (Cafferoq) [inline src/store/item-store.js]: Ah. fair enough. I'll make that change.

File: src/store/checkout-store.js (hunk 3 above)
Comment (Cafferoq) [inline src/store/checkout-store.js]: It should be caught and then failure JSON will be sent with the error message.

File: src/store/item-store.js [hunk 6]
```diff
@@ -51,6 +51,11 @@ store.registerHandler('NEW_ITEM', data => {
 store.registerHandler('NEW_CHECKOUT', data => {
     data.itemAddresses.forEach(address => {
         store.getItemByAddress(address).status = 'CHECKED_OUT';
+
+        let timestamp = new Date(data.timestamp);
```
Comment (thofmann) [inline src/store/item-store.js]: :+1:

File: test/unit/store/checkout-store.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+            return addAction('NEW_STUDENT', {
+                id: '123456',
+                name: 'John von Neumann'
+            });
+        }).then(actionId => {
+            student = StudentStore.getStudentByActionId(actionId);
+        });
+    });
+
+    it('should fail to checkout with an overdue item.', () => {
```
Comment (thofmann) [inline test/unit/store/checkout-store.js]: :+1:

**Pull request merged**